standalone-dns-proxy/cmd/standalonednsproxy.go

```diff
@@ -0,0 +1,578 @@
+				if err != nil {
+					log.WithError(err).Error("Failed to send policy state response")
+					return err
+				}
+				return err
+			}
+			response.Response = standalonednsproxy.ResponseCode_RESPONSE_CODE_NO_ERROR
+			err = sdp.policyStateStream.Send(response)
+			if err != nil {
+				log.WithError(err).Error("Failed to send policy state response")
+				return err
+			}
+		}
+	}
+}
+
+func (sdp *StandaloneDNSProxy) closePolicyStateStream() {
+	if sdp.policyStateStream != nil {
+		err := sdp.policyStateStream.CloseSend()
+		if err != nil {
+			log.Errorf("Failed to close policy state stream: %v", err)
+		}
+		sdp.policyStateStream = nil
+	}
+}
+
+func (sdp *StandaloneDNSProxy) closeConnection() error {
+	if sdp.connection != nil {
+		err := sdp.connection.Close()
+		if err != nil {
+			log.Errorf("Failed to close connection: %v", err)
+			return err
+		}
+		sdp.connection = nil
+	}
+	return nil
+}
+
+// createStreamPolicyStateStream creates a subscription stream to the policy state
+func (sdp *StandaloneDNSProxy) createStreamPolicyStateStream(ctx context.Context) error {
+	if sdp.Client == nil {
+		log.Error("Client is nil")
+		return fmt.Errorf("client is nil")
+	}
+
+	if sdp.policyStateStream != nil {
+		log.Error("Policy state stream is not nil")
+		sdp.closePolicyStateStream()
+	}
+
+	stream, err := sdp.Client.StreamPolicyState(ctx)
+	if err != nil {
+		log.WithError(err).Error("Failed to subscribe to policy state")
+		return err
+	}
+	sdp.policyStateStream = stream
+	return nil
+}
+
+// UpdatePolicyState updates the DNS rules in the standalone DNS proxy
+// 1. It updates the ip to identity cache and ip to endpoint id cache
+// 2. It updates the DNS rules in the standalone DNS proxy
+// The input is the policy state received from the cilium agent as :
+//
+//	PolicyState : {
+//	  EgressL7DnsPolicy : [
+//	    {
+//	    SourceEndpointId : 1
+//	    DnsServers : [{
+//	      DnsServerIdentity : 2
+//	      DnsServerPort : 53
+//	      DnsServerProto : 17
+//	    	},
+//	    	{
+//	      DnsServerIdentity : 3
+//	      DnsServerPort : 53
+//	      DnsServerProto : 17
+//	    	}]
+//	    DnsPattern : ["www.example.com"]
+//	    },
+//	    {
+//	    SourceEndpointId : 1
+//	    DnsServers : [{
+//	      DnsServerIdentity : 2
+//	      DnsServerPort : 54
+//	      DnsServerProto : 17
+//	    	}]
+//	    DnsPattern : ["www.test.com"]
+//	    }
+//	  ]
+//	  IdentityToEndpointMapping : []
+//	}
+func (sdp *StandaloneDNSProxy) UpdatePolicyState(rules *standalonednsproxy.PolicyState) (revert.RevertStack, error) {
+	log.Debugf("Received policy state: %v", rules)
+
+	var revertStack revert.RevertStack
+	identityToEndpointMapping := rules.GetIdentityToEndpointMapping()
+
+	// Update the ip to identity cache and ip to endpoint id cache
+	for _, mapping := range identityToEndpointMapping {
+		for _, epInfo := range mapping.GetEndpointInfo() {
+			for _, ip := range epInfo.GetIp() {
+				sdp.ipToIdentityCache[string(ip)] = mapping.GetIdentity()
```

Comment:
I think this write and the above mentioned read need to be synchronised, since writing and reading from a map is not safe in Go.

Reply:
Yes, you are right. Update it to use the lock to make sure they are safe.